File: 프로그래머스/2/92342. 양궁대회/양궁대회.py
from itertools import combinations_with_replacement


def get_ryan_board(scores):
    result = [0] * 11
    
    for score in scores:
        result[10 - score] += 1
    
    return result


def get_scores(apeach_board, ryan_board):
    apeach_score = ryan_score = 0
    
    for i, (a, b) in enumerate(zip(apeach_board, ryan_board)):
        score = 10 - i
        
        if a == b == 0:
            continue
        
        if a >= b:
            apeach_score += score
        else:
            ryan_score += score
    
    return apeach_score, ryan_score


def solution(n, info):
    answer = [-1]
    max_gap = -1
    
    for shoot in combinations_with_replacement(range(11), n):
        ryan_board = get_ryan_board(shoot)
        apeach_score, ryan_score = get_scores(info, ryan_board)
    
        if ryan_score <= apeach_score:
            continue
        
        if ryan_score - apeach_score > max_gap:
            max_gap = ryan_score - apeach_score
            answer = ryan_board
    
    return answer

File: 프로그래머스/2/92342. 양궁대회/양궁대회.java
import java.util.*;

class Solution {

    private int maxGap = -1;
    private int[] answer = {-1};

    public int[] solution(int n, int[] info) {
        int[] ryan = new int[11];
        dfs(n, 0, ryan, info);
        return answer;
    }

    private void dfs(int remain, int start, int[] ryan, int[] info) {
        if (remain == 0) {
            int[] candidate = Arrays.copyOf(ryan, 11);
            int[] scores = getScores(info, candidate);

            int apeach = scores[0];
            int ryanScore = scores[1];

            if (ryanScore > apeach) {
                int gap = ryanScore - apeach;

                if (gap > maxGap) {
                    maxGap = gap;
                    answer = candidate;
                } else if (gap == maxGap) {
                    if (isBetter(candidate, answer)) {
                        answer = candidate;
                    }
                }
            }
            return;
        }

        for (int i = start; i < 11; i++) {
            ryan[i]++;
            dfs(remain - 1, i, ryan, info);
            ryan[i]--;
        }
    }

    private int[] getScores(int[] apeachBoard, int[] ryanBoard) {
        int apeachScore = 0;
        int ryanScore = 0;

        for (int i = 0; i < 11; i++) {
            int score = 10 - i;

            if (apeachBoard[i] == 0 && ryanBoard[i] == 0) continue;

            if (apeachBoard[i] >= ryanBoard[i]) {
                apeachScore += score;
            } else {
                ryanScore += score;
            }
        }

        return new int[]{apeachScore, ryanScore};
    }

    private boolean isBetter(int[] a, int[] b) {
        for (int i = 10; i >= 0; i--) {
            if (a[i] != b[i]) {
                return a[i] > b[i];
            }
        }
        return false;
    }
}